panel/custom.py

```diff
@@ -0,0 +1,387 @@
+                    continue
+                old._cleanup(root)
+        if any(e in model.children for e in events):
+            children = self._get_children(model.data, doc, root, model, comm)
+            data_msg.update(children)
+            model_msg['children'] = list(children)
+        self._set_on_model(model_msg, root, model)
+        self._set_on_model(data_msg, root, model.data)
+
+    def on_event(self, event: str, callback: Callable) -> None:
+        """
+        Registers a callback to be executed when the specified DOM
+        event is triggered.
+
+        Arguments
+        ---------
+        event: str
+          Name of the DOM event to add an event listener to.
+        callback: callable
+          A callable which will be given the DOMEvent object.
+        """
+        self._event_callbacks[event].append(callback)
+
+
+class JSComponent(ReactiveESM):
+    '''
+    The `JSComponent` allows you to create custom Panel components
+    using Javascript and CSS without the complexities of
+    Javascript build tools.
+
+    A `JSComponent` subclass provides bi-directional syncing of its
+    parameters with arbitrary HTML elements, attributes and
+    properties. The key part of the subclass is the `_esm`
+    variable. Use this to define a `render` function as shown in the
+    example below.
+
+    import panel as pn
+    import param
+
+    pn.extension()
+
+    class CounterButton(JSComponent):
+
+        value = param.Integer()
+
+        _esm = """
+        export function render({ data }) {
+            let btn = document.createElement("button");
+            btn.innerHTML = `count is ${data.value}`;
+            btn.addEventListener("click", () => {
+                data.value += 1
+            });
+            data.watch(() => {
+                btn.innerHTML = `count is ${data.value}`;
+            }, 'value')
+            return btn
+        }
+        """
+
+    CounterButton().servable()
+    '''
+
+
+class ReactComponent(ReactiveESM):
+    '''
+    The `ReactComponent` allows you to create custom Panel components
+    using React without the complexities of Javascript build tools.
+
+    A `ReactComponent` subclass provides bi-directional syncing of its
+    parameters with arbitrary HTML elements, attributes and
+    properties. The key part of the subclass is the `_esm`
+    variable. Use this to define a `render` function as shown in the
+    example below.
+
+    import panel as pn
+    import param
+
+    pn.extension()
+
+    class CounterButton(ReactComponent):
+
+        value = param.Integer()
+
+        _esm = """
+        export function render({ data, state }) {
+          return (
+            <>
+              <button onClick={() => { data.value += 1 }}>{state.value}</button>
+            </>
+          )
+        }
+        """
+
+    CounterButton().servable()
+    '''
+
+
+    _bokeh_model = _BkReactComponent
+
+    _react_version = '18.2.0'
```

Comment:
Would be helpful if development version of react was used when using `--autoreload`.

For example right now I'm seeing

![image](https://github.com/holoviz/panel/assets/42288570/88b65af2-3602-40ee-a66c-13a999505200)

Thus I'm stuck as I cannot get error message.

```python
import panel as pn
from panel.custom import ReactComponent, Children
from panel.layout.base import NamedListLike

pn.extension()


class ListLikeLayout(NamedListLike, ReactComponent):
    objects = Children()

    _esm = """
    export function Render({ model }) {
        const objects = model.get_child("objects");

        return (
            <div>
                {objects.map((object, index) => (
                    <React.Fragment key={index}>
                        {object}
                        {index < objects.length - 1 && <div className="divider"></div>}
                    </React.Fragment>
                ))}
            </div>
        );
    }"""

    _stylesheets = [
        """
.divider {border-top: 3px dotted #bbb}; 
"""
    ]


layout = ListLikeLayout(
    "I love beat boxing",
    "https://upload.wikimedia.org/wikipedia/commons/d/d3/Beatboxset1_pepouni.ogg",
    "Yes I do!",
    styles={"border": "2px solid lightgray"},
).servable()
```

Reply:
Yes, indeed, was planning to improve the React error reporting.

Reply:
Also done!